pkg/reconciler/serverlessservice/resources/services.go

```diff
@@ -33,16 +33,39 @@ const (
 	servicePortNameH2C   = "http2"
 
 	// ServicePort is the external port of the service
-	servicePort = int32(80)
+	servicePort              = int32(80)
+	serviceHTTP2Port         = int32(81)
+	activatorTargetHTTP1Port = 8080
+	activatorTargetHTTP2Port = 8081
 
 	// RequestQueuePortName specifies the port name to use for http requests
 	// in queue-proxy container.
 	requestQueuePortName string = "queue-port"
 )
 
+func pubTargetPort(sks *v1alpha1.ServerlessService, hasBackends bool) intstr.IntOrString {
+	switch {
+	case sks.Spec.Mode == v1alpha1.SKSOperationModeServe && hasBackends:
+		return intstr.FromString(requestQueuePortName)
+	case sks.Spec.Mode == v1alpha1.SKSOperationModeProxy || !hasBackends:
+		if sks.Spec.ProtocolType == networking.ProtocolH2C {
+			return intstr.FromInt(activatorTargetHTTP2Port)
+		}
+		return intstr.FromInt(activatorTargetHTTP1Port)
+	}
+	panic("webhook validation failed")
```

Comment:
given `hasBackends` argument, I don't completely buy this...

Reply:
eh, if truth table is 2x2 (proxy/serve x true/false) the switch covers all 4 cells. So the only possibility is hasBackends == true and mode is third value.